firmware/src/msg.c

```diff
@@ -91,6 +93,31 @@ static enum bl_error bl_msg_source_cap(
 			msg->source_cap_req.source, &response->source_cap);
 }
 
+static enum bl_error bl_msg_version(
+		const union bl_msg_data *msg,
+		union bl_msg_data *response)
+{
+	/* commit sha needs to be interpreted as a string then stored in a buffer to be interpreted as hex */
+	char commit_sha[COMMIT_SHA_LENGTH * 8] = BL_COMMIT_SHA;
+	char buf[9] = {0}; /* 8 characters and a null-terminator */
+	int ret;
+	uint32_t value;
+	BL_UNUSED(msg);
+	response->version.type = BL_MSG_VERSION;
+	response->version.revision = BL_REVISION;
+	for (unsigned i = 0; i < COMMIT_SHA_LENGTH; i++) {
+		strncpy(buf, commit_sha + i * 8, 8);
+		ret = sscanf(buf, "%"PRIx32, &value);
```

Comment:
This function looks good and it should work fine! I've got a few comments about how it might be done more optimally, in case you're interested, but really this isn't performance critical code, so it's fine as is.

It should be possible to avoid the copy into buff. Maybe try setting the max input character width in the scanf format specifier: `%8"PRIx32`.  Also, it should be doable to read straight from the string literal `BL_COMMIT_SHA`, avoiding the need to copy to `commit_sha` on the stack.  One way would be to do the for loop as `(unsigned i = 0; i < sizeof(BL_COMMIT_SHA) - 1; i += 8)` and read from `&BL_COMMIT_SHA[i]`.

Obviously you depend on the `BL_COMMIT_SHA` being of the correct length. C11 added `_Static_assert` for doing these kinds of compile time checks, and traditionally the more portable way is to define a macro like [I did here](https://github.com/tlsa/libcyaml/blob/master/src/util.h#L18). So you could have something like:

```c
bl_static_assert(sizeof(BL_COMMIT_SHA) - 1 == COMMIT_SHA_LENGTH * 8);
```

Reply:
Another thing is that sscanf is quite a fat function to put on a microcontroller.  Another way to encode it into the uint32_t data would be something like:

```c
for (unsigned i = 0; i < 5; i++) {
	response->version.commit_sha[i] = 0;
	for (unsigned j = 0; j < 8; j++) {
		response->version.commit_sha[i] |=
				hex_to_val(BL_COMMIT_SHA[i * 8 + j]) << 4 * j;
	}
}
```

(Each hex char represents 4 bits in binary.)

Reply:
Another method still would be to have compile time tooling generate 5 uint32_t values at build time, and bake those into the firmware.  Maybe something like `git rev-parse --verify HEAD | sed 's/.\{8\}/0x&,/g'` and then:

```c
const uint32_t bl_build_sha[5] = { BL_COMMIT_SHA };
```

Reply:
Oh, also, it should be `SCNx32` for the `scanf` functions.

Reply:
When I use `hex_to_val(BL_COMMIT_SHA[i * 8 + j]) << 4 * j;` I end up with the order of digits reversed, however it works with `7 - j` instead of `j`. Is there a better way of doing it?

Reply:
That way is fine. The alternative would be to manually write the decoder on the other end.